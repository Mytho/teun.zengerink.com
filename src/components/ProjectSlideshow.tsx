import { useRouter, NextRouter } from 'next/router'
import ReactMarkdown from 'react-markdown'
import { Project, Photo } from '../projectService'
import styles from './ProjectSlideshow.module.scss'

interface ProjectSlideshowProps {
  project: Project
}

class ProjectRouter {
  private router: NextRouter
  private project: Project

  constructor(router: NextRouter, project: Project) {
    this.router = router
    this.project = project
  }

  private getSlideIndex(): number {
    return this.project.photos.findIndex((photo) => photo.key === this.activeKey())
  }

  private firstPhoto(): Photo {
    return this.project.photos[0]
  }

  private lastPhoto(): Photo {
    return [...this.project.photos].pop()
  }

  private nextPhoto(): Photo {
    return this.project.photos[this.getSlideIndex() + 1]
  }

  private previousPhoto(): Photo {
    return this.project.photos[this.getSlideIndex() - 1]
  }

  private isLast(): boolean {
    return this.activeKey() === this.lastPhoto().key
  }

  private isFirst(): boolean {
    return this.activeKey() === this.firstPhoto().key
  }

  private go(photo?: Photo): Promise<boolean> {
    const slug = `/work/${this.project.slug}`
    if (!photo) return this.router.push(slug)
    return this.router.push(`${slug}/${photo.key}`)
  }

  isActive(photo: Photo): boolean {
    return this.activeKey() === photo.key
  }

  activeKey() {
    return this.router.query?.slug ? this.router.query?.slug[1] : undefined
  }

  next() {
    if (this.project.statement) {
      if (!this.activeKey()) return this.go(this.firstPhoto())
      if (this.isLast()) return this.go()
      return this.go(this.nextPhoto())
    }

    if (this.isLast()) return this.go(this.firstPhoto())
    this.go(this.nextPhoto())
  }

  previous() {
    if (this.project.statement) {
      if (!this.activeKey()) return this.go(this.lastPhoto())
      if (this.isFirst()) return this.go()
      return this.go(this.previousPhoto())
    }

    if (this.isFirst()) return this.go(this.lastPhoto())
    this.go(this.previousPhoto())
  }
}

const ProjectSlideshow = (props: ProjectSlideshowProps): React.ReactElement => {
  const router = new ProjectRouter(useRouter(), props.project)

  return (
    <div className={styles.slideshow}>
      {props.project?.statement ? (
        <div
          className={`${styles.slide} ${router.activeKey() === undefined ? styles.active : ''}`}
          onClick={() => router.next()}
        >
          <h2>{props.project?.title}</h2>
          <ReactMarkdown source={props.project?.statement} />
        </div>
      ) : (
        ''
      )}
      {props.project?.photos.map((photo) => (
        <img
          key={`${props.project.slug}--photo-${photo.key}`}
          className={`${styles.slide} ${router.isActive(photo) ? styles.active : ''}`}
          src={photo.url}
          alt={photo.caption}
          onClick={() => router.next()}
        />
      ))}
      <div className={styles.navigation}>
        <a className={styles.leftArrow} onClick={() => router.previous()}>
          &larr;
        </a>
        <h3>{router.activeKey()}</h3>
        <a className={styles.rightArrow} onClick={() => router.next()}>
          &rarr;
        </a>
      </div>
      {props.project?.photos
        .filter((photo) => photo.caption)
        .map((photo) => (
          <div
            key={`${props.project.slug}--caption-${photo.key}`}
            className={`${styles.caption} ${router.isActive(photo) ? styles.active : ''}`}
          >
            <ReactMarkdown source={photo.caption} />
          </div>
        ))}
    </div>
  )
}

export default ProjectSlideshow
