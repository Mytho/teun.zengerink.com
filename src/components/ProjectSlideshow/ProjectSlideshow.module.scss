@import '../../styles/variables.scss';

.slideshow {
  position: relative;
  display: inline-block;
}

.slide {
  display: none;
  text-align: center;
  font-size: 1em;
  line-height: 0;
}

.previous,
.next {
  position: absolute;
  top: 0;
  bottom: 0;
}

.previous {
  left: 0;
  right: 50%;
  cursor: w-resize;
}

.next {
  left: 50%;
  right: 0;
  cursor: e-resize;
}

.active {
  display: inline-block;
}

@media (min-width: $phone-breakpoint) and (max-width: $tablet-portrait-breakpoint) {
  .slideshow {
    margin-top: $outer-margin;
  }
}

@media (min-width: $tablet-portrait-breakpoint + 1) {
  .slideshow {
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
  }
}
