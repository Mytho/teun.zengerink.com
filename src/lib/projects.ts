import fs from 'fs'
import path from 'path'
import slugify from 'slugify'
import sizeOf from 'image-size'
import yaml from 'js-yaml'

interface ConfigCaption {
  key: string
  caption: string
}

interface ConfigItem {
  title: string
  statement?: string
  captions?: ConfigCaption[]
}

export interface Photo {
  key: string
  url: string
  size: { width: number; height: number }
  caption?: string
}

export interface Project {
  title: string
  slug: string
  photos: Photo[]
  statement?: string
}

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), 'public/photos', slug)
  const fileNames = fs.readdirSync(photosDirectory)

  return fileNames.map((filename) => {
    const key = path.basename(filename, '.jpg')
    const caption = captions?.find((c) => c.key == key)

    return {
      key,
      caption: caption?.caption ?? '',
      url: `/photos/${slug}/${filename}`,
      size: sizeOf(path.join(photosDirectory, filename)),
    }
  })
}

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true })

    return {
      slug,
      title: project.title,
      statement: project.statement ?? '',
      photos: generatePhotos(slug, project.captions),
    }
  })
}

export const getProjects = async (): Promise<Project[]> => {
  try {
    const file = path.join(process.cwd(), 'src', 'config.yml')
    const config = yaml.load(fs.readFileSync(file, 'utf-8'))
    return await generateProjects(config)
  } catch (e) {
    console.error(e)
  }
}
