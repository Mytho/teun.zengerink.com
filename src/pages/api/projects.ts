import fs from "fs";
import { IncomingMessage, ServerResponse } from 'http'
import path from "path";
import slugify from "slugify";

interface ConfigCaption {
  key: string;
  caption: string;
}

interface ConfigItem {
  title: string;
  statement?: string;
  captions?: ConfigCaption[];
}

interface Photo {
  key: string;
  url: string;
  caption?: string;
}

export interface Project {
  title: string;
  slug: string;
  photos: Photo[];
  statement?: string;
}

const generatePhotos = (slug: string, captions: ConfigCaption[]): Photo[] => {
  const photosDirectory = path.join(process.cwd(), "public/photos", slug);
  const fileNames = fs.readdirSync(photosDirectory);

  return fileNames.map((filename) => {
    const key = path.basename(filename, ".jpg");
    const caption = captions?.find((c) => c.key == key)?.caption;

    return {
      key,
      caption,
      url: `photos/${slug}/${filename}`,
    };
  });
};

const generateProjects = (item: ConfigItem[]): Project[] => {
  return item.map((project) => {
    const slug = slugify(project.title, { lower: true });

    return {
      title: project.title,
      statement: project.statement,
      slug,
      photos: generatePhotos(slug, project.captions),
    };
  });
};

const config: ConfigItem[] = [
  {
    title: "Take, Make, Waste",
    statement: "TODO",
    captions: [
      {
        key: "01",
        caption: "Image 01: TODO",
      },
      {
        key: "02",
        caption: "Image 02: TODO",
      },
      {
        key: "03",
        caption: "Image 03: TODO",
      },
    ],
  },
  {
    title: "Point of no Return",
    statement: "TODO",
  },
  {
    title: "Under Your Nose",
    statement: "TODO",
  },
  {
    title: "Fortress",
    statement: "TODO",
  },
  {
    title: "Symbiosis",
    statement: "TODO",
  },
  {
    title: "Untitled",
  },
];

const handler = (request: IncomingMessage, response: ServerResponse): void => {
  response.statusCode = 200;
  response.setHeader("Content-Type", "application/json");
  response.end(JSON.stringify({ projects: generateProjects(config) }));
};

export default handler;
