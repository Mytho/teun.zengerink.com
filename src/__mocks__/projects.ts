import { Project } from '../lib/projects'

const projects: Project[] = [
  {
    title: 'My First Project',
    slug: 'my-first-project',
    statement: 'The statement for the project',
    photos: [
      {
        key: '01',
        url: '/photos/my-first-project/01.jpg',
        size: { width: 800, height: 600 },
      },
      {
        key: '02',
        url: '/photos/my-first-project/02.jpg',
        size: { width: 600, height: 800 },
      },
      {
        key: '03',
        url: '/photos/my-first-project/03.jpg',
        size: { width: 800, height: 600 },
      },
    ],
  },
  {
    title: 'Project without a statement',
    slug: 'no-statement',
    photos: [
      {
        key: '01',
        url: '/photos/no-statement/01.jpg',
        caption: 'This is image 1',
        size: { width: 800, height: 600 },
      },
      {
        key: '02',
        url: '/photos/no-statement/02.jpg',
        caption: 'This is image 2',
        size: { width: 800, height: 600 },
      },
      {
        key: '03',
        url: '/photos/no-statement/03.jpg',
        caption: 'This is image 3',
        size: { width: 600, height: 800 },
      },
      {
        key: '04',
        url: '/photos/no-statement/04.jpg',
        caption: 'This is image 4',
        size: { width: 800, height: 600 },
      },
    ],
  },
]

export default projects
