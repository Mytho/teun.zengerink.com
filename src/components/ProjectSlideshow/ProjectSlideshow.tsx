import React, { useEffect } from 'react'
import ReactMarkdown from 'react-markdown'
import { Project } from '../../lib/projects'
import { useProject } from '../../lib/useProject'
import Photo from '../Photo/Photo'
import styles from './ProjectSlideshow.module.scss'

enum Key {
  LEFT = 'ArrowLeft',
  RIGHT = 'ArrowRight',
}

interface ProjectSlideshowProps {
  project: Project
}

const pad = (str: string, length: number): string => {
  const output = str?.toString() ?? ''
  return output.length < length ? pad('0' + output, length) : output
}

const ProjectSlideshow: React.FC<ProjectSlideshowProps> = ({ project }) => {
  let touchStartX: number
  const { activeKey, previous, next } = useProject(project)

  const numberOfPhotos = pad(project?.photos?.length?.toString(), 2)

  useEffect(() => {
    const keyUpHandler = ({ key }: KeyboardEvent) => {
      switch (key) {
        case Key.LEFT:
          previous()
          break
        case Key.RIGHT:
          next()
          break
      }
    }

    const ON = 'keyup'
    window.addEventListener(ON, keyUpHandler)
    return () => window.removeEventListener(ON, keyUpHandler)
  })

  const touchStartHandler = ({ touches }: React.TouchEvent) => {
    touchStartX = touches[0].clientX
  }

  const touchMoveHandler = ({ touches }: React.TouchEvent) => {
    if (!touchStartX) return
    const difference = touchStartX - touches[0].clientX
    if (difference < 0) previous()
    if (difference > 0) next()
    touchStartX = null
  }

  return (
    <div className={styles.slideshow}>
      {project?.statement && (
        <div
          className={`${styles.slide} ${activeKey === undefined ? styles.active : ''}`}
          onClick={() => next()}
          onTouchStart={(e) => touchStartHandler(e)}
          onTouchMove={(e) => touchMoveHandler(e)}
        >
          <h2>{project?.title}</h2>
          <ReactMarkdown>{project?.statement}</ReactMarkdown>
        </div>
      )}
      {project?.photos.map((photo) => (
        <div
          key={`${project.slug}--slide-${photo.key}`}
          className={`${styles.slide} ${photo.key === activeKey ? styles.active : ''}`}
          onClick={() => next()}
          onTouchStart={(e) => touchStartHandler(e)}
          onTouchMove={(e) => touchMoveHandler(e)}
        >
          <Photo photo={photo} alt={`${project?.title} - ${photo.key}`} />
        </div>
      ))}
      <div className={styles.navigation}>
        <a className={styles.leftArrow} onClick={() => previous()}>
          &larr;
        </a>
        {activeKey && (
          <h3>
            {activeKey}/{numberOfPhotos}
          </h3>
        )}
        <a className={styles.rightArrow} onClick={() => next()}>
          &rarr;
        </a>
      </div>
      {project?.photos
        .filter((photo) => photo.caption)
        .map((photo) => (
          <div
            key={`${project.slug}--caption-${photo.key}`}
            className={`${styles.caption} ${photo.key === activeKey ? styles.active : ''}`}
          >
            <ReactMarkdown>{photo.caption}</ReactMarkdown>
          </div>
        ))}
    </div>
  )
}

export default ProjectSlideshow
