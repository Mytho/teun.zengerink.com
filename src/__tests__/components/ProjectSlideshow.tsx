import React from 'react'
import renderer from 'react-test-renderer'
import ProjectSlideshow from '../../components/ProjectSlideshow'
import projects from '../../__mocks__/projects'

const defaultProps = { project: projects[0] }

it('renders correctly', () => {
  const tree = renderer.create(<ProjectSlideshow {...defaultProps} />).toJSON()
  expect(tree).toMatchSnapshot()
})
