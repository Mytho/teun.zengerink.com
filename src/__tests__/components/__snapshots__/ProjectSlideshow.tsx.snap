// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="slideshow"
>
  <div
    className="slide active"
    onClick={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
  >
    <h2>
      My First Project
    </h2>
    <p>
      The statement for the project
    </p>
  </div>
  <div
    className="slide "
    onClick={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
  >
    <img
      alt="01"
      height={600}
      src="/photos/my-first-project/01.jpg"
      width={800}
    />
  </div>
  <div
    className="slide "
    onClick={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
  >
    <img
      alt="02"
      height={800}
      src="/photos/my-first-project/02.jpg"
      width={600}
    />
  </div>
  <div
    className="slide "
    onClick={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
  >
    <img
      alt="03"
      height={600}
      src="/photos/my-first-project/03.jpg"
      width={800}
    />
  </div>
  <div
    className="navigation"
  >
    <a
      className="leftArrow"
      onClick={[Function]}
    >
      ←
    </a>
    
    <a
      className="rightArrow"
      onClick={[Function]}
    >
      →
    </a>
  </div>
</div>
`;
